0.3.0 release. (#99)
js-temporal/temporal-polyfill: CHANGELOG.md

# 0.3.0

This version roughly corresponds with all the changes made to the Temporal polyfill as of the October 2021 TC39 Plenary meeting.

Breaking changes:

> - Timezones now require a `getOffsetNanosecondsFor` method, and no longer fall back to the intrinsic definition (previously provided by `Temporal.Timezone#getOffsetNanosecondsFor`) if not provided. ([08346dc5])

Reviewer: Is there a corresponding proposal-temporal PR? OK if not, just wanted to make sure.

Author: There is a corresponding PR, but weirdly it doesn't change the spec text. Seems maybe the spec text was already  this way (for a while): https://github.com/tc39/proposal-temporal/blame/60c753a6369c0a61cacc353562214aaa6327117a/spec/timezone.html#L586-L597

- Allow ISO strings with "Z" + a bracketed IANA name ([70bd9898], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1749))
- Emit ES2020 builds for newer browsers, and emit ES5 for older browsers. ([2331468d], [9e95c62b])
- Temporal.Duration constructor will now throw if given a non-integer ([9df5d068], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1872))
- Remove support for sub-minute offsets in ISO strings ([766e5037], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1871), [Spec PR](https://github.com/tc39/proposal-temporal/pull/1862))
- Throw TypeError on missing options from Duration.total ([4ec075f0], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1720))
- Reject non-integer Duration fields in Duration.with() ([e6b2488d], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1735))
- Ensure an Object is returned from calendar.mergeFields() ([4e63f25f], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1719))

Bug fixes:

- Fix GetFormatterParts for Firefox Nightly ([47f9132f])
- Fix TS types of RoundTo and TotalOf ([3008a670])

> - Fix crash setting `day` outside current JPN era ([6d3588c3], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1807))

Reviewer: Minor nit: I don't think this was a spec change, so "Spec PR" isn't the right title for the link. Maybe "Related PR" ?  "Related proposal-temporal PR" ?  Or any other label you think would make sense.

Author: Changed all the labels where the spec isn't changed to be "proposal-temporal polyfill PR".

- Copy options object for PlainYearMonth.{add,subtract} and InterpretTemporalDateTimeFields to prevent user-modified objects from interfering with later operations. ([bafa1bdf], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1748))
- Validate input to Calendar.prototype.fields ([7ebc700e], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1750))
- Stop observably calling into `getPossibleInstantsFor` in `InterpretISODateTimeOffset` ([5448e59f], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1688))
- Call `CalendarEquals` correctly ([07ea694e], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1858))

> - Fix arithmetic issues when using non-ISO months ([079a3325], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1761))

Reviewer: Not a spec change.

- Regex: tighten matching of month and day values in datesplit ([b5736546], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1836))

> - Fix TS types for required CalendarProtocol methods ([0ee4581f], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1964))

Reviewer: Same here. Not a spec change.

# 0.3.0

> This version roughly corresponds with all the changes made to the Temporal polyfill as of the November TC39 Plenary meeting.

Reviewer: ```suggestion
This version roughly corresponds with all the changes made to the Temporal polyfill as of the November 2021 TC39 Plenary meeting.
```

Reviewer 2: There wasn't a November meeting - I think this should be October :smile:

Author: Oops - Fixed :D

> Backwards-incompatible changes:

Reviewer: Would "Breaking Changes" be clearer?

Author: Yep!

>  - Timezones now require a `getOffsetNanosecondsFor` method, and no longer fall back to the intrinsic definition (previously provided by `Temporal.Timezone#getOffsetNanosecondsFor`) if not provided. 08346dc5

Reviewer: I think commit IDs in MD files need to be HTTPS links, otherwise they'll show up as plain text to the reader.

Reviewer: BTW I'm not sure about this; in just not seeing the commits show up as links when editing here on my phone.

Author: Unfortunately you are right:  https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/autolinked-references-and-urls notes that this doesn't happen for wikis or files within repos.

Author: I've added all the commit links, and added some notes in the release-checklist on how to quickly regenerate the list of links.

 - Emit ES2020 builds for newer browsers, and emit ES5 for older browsers. (2331468d, 9e95c62b)
 - Temporal.Duration constructor will now throw if given a non-integer (9df5d068)
 - Remove support for sub-minute offsets in ISO strings (766e5037, see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1871))
 - Throw TypeError on missing options from Duration.total (4ec075f0)
 - Reject non-integer Duration fields in Duration.with() (e6b2488d)
 - Ensure an Object is returned from calendar.mergeFields() (4e63f25f)

Bug fixes:
 - Fix GetFormatterParts for Firefox Nightly (47f9132f)
 - Fix TS types of RoundTo and TotalOf (3008a670)
 - Fix crash setting `day` outside current JPN era (6d3588c3)
 - Copy options object for PlainYearMonth.{add,subtract} and InterpretTemporalDateTimeFields to prevent user-modified objects from interfering with later operations. (bafa1bdf)
 - Validate input to Calendar.prototype.fields (7ebc700e)
 - Stop observably calling into `getPossibleInstantsFor` in `InterpretISODateTimeOffset` (5448e59f)
 - Call `CalendarEquals` correctly (07ea694e)
 - Fix arithmetic issues when using non-ISO months (079a3325)

New changes:

> - Various `#round` and `#total` methods now accept string parameters or options bags. Strings are interpreted as the `smallestUnit` argument. (068e801f)

Reviewer: ```suggestion
- Various `#round` and `#total` methods now accept string parameters or options bags. Strings are interpreted as the `smallestUnit` option (or `unit` for `Temporal. Duration#total`). (068e801f)
```

 - Add @@toStringTag to TS types (41ab6bc0)
 - Accept string Calendar names in PlainMonthDay and PlainYearMonth constructors (27b4c7e8)
 - Make options optional in Calendar method TS types (3a09d00d)
 - Regex: tighten matching of month and day values in datesplit (b5736546)

Other:
 - Bump various dependencies (47701107, f5427de9, 310d9d8b)
 - Allow launching and debugging tests (both Demitasse and Test262 suites) from the VSCode debug panel (960d9b76, 7f7c19a1, 4ec6568e, edcc668b)
 - Run the Test262 test suite against this polyfill, for various configurations of the resulting build artifact (2331468d, 666c69da, 429273ec, ff937782, f885253f)
 - Remove various pieces of unused code, and add CI testing to detect unused code in PR review (67f9f6bb, 63bdfcd1)
 - Drop the dependency on es-abstract (d24575f2, ad7e2e3a, 5b1bc5e2)

>  - The polyfill's source has move to TypeScript (12e4d529, ac78fd9d, 53f32e0f, 06b806c9, 66fdc765, 50b1c34b, 4724b017, 947a8a5e, fdbf7e01, fa60af6a, da753f2f, f4db8b0b, 4a38420d)

Reviewer: ```suggestion
 - The polyfill's source was ported to TypeScript (12e4d529, ac78fd9d, 53f32e0f, 06b806c9, 66fdc765, 50b1c34b, 4724b017, 947a8a5e, fdbf7e01, fa60af6a, da753f2f, f4db8b0b, 4a38420d)
```

Author: Done.

Bug fixes:

- Fix GetFormatterParts for Firefox Nightly ([47f9132f])
- Fix TS types of RoundTo and TotalOf ([3008a670])
- Fix crash setting `day` outside current JPN era ([6d3588c3], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1807))
- Copy options object for PlainYearMonth.{add,subtract} and InterpretTemporalDateTimeFields to prevent user-modified objects from interfering with later operations. ([bafa1bdf], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1748))
- Validate input to Calendar.prototype.fields ([7ebc700e], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1750))
- Stop observably calling into `getPossibleInstantsFor` in `InterpretISODateTimeOffset` ([5448e59f], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1688))
- Call `CalendarEquals` correctly ([07ea694e], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1858))
- Fix arithmetic issues when using non-ISO months ([079a3325], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1761))

> New changes:

Reviewer: Should we call this "Non-breaking changes" for symmetry, and move this section below the "Breaking changes" section?

Author: Maybe call it "New features", but keep it where it is?

 Personally, I think "breaking changes > bug fixes > new stuff that isn't a bugfix > misc documentation for maintainers" is the right order.

Reviewer: > Maybe call it "New features", but keep it where it is?
For a regular OSS package I'd agree, but as a Stage 3 polyfill claiming that there are new "features" seems like a bad idea because the idea of Stage 3 is that there won't be any new features!  I'd prefer to call them "Non-breaking changes" or some other wording that doesn't make it sound like the proposal is a moving target.

> Personally, I think "breaking changes > bug fixes > new stuff that isn't a bugfix > misc documentation for maintainers" is the right order.

Sure that's OK with me.

Author: I suppose in terms of functionality there's no new features, though accepting strings to round and total sure seemed like a "feature" to me 😆 

Non-breaking definitely makes more sense with your explanation, so let's use that.

- Various `#round` and `#total` methods now accept string parameters or options bags. Strings are interpreted as the `smallestUnit` option (or `unit` for `Temporal.Duration#total`). ([068e801f], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1875))
- Add @@toStringTag to TS types ([41ab6bc0])
- Accept string Calendar names in PlainMonthDay and PlainYearMonth constructors ([27b4c7e8])
- Make options optional in Calendar method TS types ([3a09d00d])

> - Regex: tighten matching of month and day values in datesplit ([b5736546], see also [Spec PR](https://github.com/tc39/proposal-temporal/pull/1836))

Reviewer: Is this really a non-breaking change?  Given that it introduces new exceptions for cases that didn't throw before, is it a bug fix or a breaking change?

Author: Agreed this is actually a bug fix, not a non-breaking change. Moved.

Other:

- Bump various dependencies ([47701107], [f5427de9], [310d9d8b])
- Allow launching and debugging tests (both Demitasse and Test262 suites) from the VSCode debug panel ([960d9b76], [7f7c19a1], [4ec6568e], [edcc668b])
- Run the Test262 test suite against this polyfill, for various configurations of the resulting build artifact ([2331468d], [666c69da], [429273ec], [ff937782], [f885253f])
- Remove various pieces of unused code, and add CI testing to detect unused code in PR review ([67f9f6bb], [63bdfcd1])
- Drop the dependency on es-abstract ([d24575f2], [ad7e2e3a], [5b1bc5e2])
- The polyfill's source was ported to TypeScript ([12e4d529], [ac78fd9d], [53f32e0f], [06b806c9], [66fdc765], [50b1c34b], [4724b017], [947a8a5e], [fdbf7e01], [fa60af6a], [da753f2f], [f4db8b0b], [4a38420d])

> - Document the release process for this polyfill ([c55818b6])

Reviewer: This is so cool. Thanks for doing this!

Author: Of course! I hope to automate this process just a little more in the future so we can always be certain this changelog makes sense.

Something that would help greatly I think going forward would be some sort of check during PR review to either tag a PR with the relevant spec PRs, or to explicitly acknowledge that there are no corresponding PRs. I'm going to ponder this a bit over the holidays and might throw something together for us to consider in the new year.
